polkadot/node/network/statement-distribution/src/v2/mod.rs

```diff
@@ -3174,6 +3181,62 @@ pub(crate) async fn handle_response<Context>(
 					"Successfully received candidate"
 				);
 
+				if !per_session.candidate_receipt_v2_enabled() &&
+					candidate.descriptor.version() == CandidateDescriptorVersion::V2
+				{
+					gum::debug!(
+						target: LOG_TARGET,
+						?candidate_hash,
+						?peer,
+						"Version 2 candidate receipts are not enabled by the runtime"
+					);
+
+					// Punish peer.
+					modify_reputation(
+						reputation,
+						ctx.sender(),
+						peer,
+						COST_INVALID_DESCRIPTOR_VERSION,
+					)
+					.await;
+					return
+				}
+
+				// Get the latest session index & check candidate descriptor session index.
+				match (candidate.descriptor.session_index(), state.per_session.last_key_value()) {
+					(Some(session_index), Some((latest_session_index, _))) => {
+						if &session_index != latest_session_index {
+							// Punish peer.
+							modify_reputation(
+								reputation,
+								ctx.sender(),
+								peer,
+								COST_INVALID_SESSION_INDEX,
+							)
+							.await;
+							return
+						}
+						// TODO: determine if we need to buffer candidates at session boundaries.
+					},
+					_ => {},
+				}
+
+				// Validate the core index.
+				if let Err(err) = candidate.check_core_index(&relay_parent_state.transposed_cq) {
```

Comment:
It would be ideal for `check_core_index` to also checks if UMP signal is malformed and return an error in that case.

also, naming change in primitives: `CandidateReceiptError` -> `CandidateCommittedReceiptError`